From: Sergio Costas <rastersoft@gmail.com>
Date: Sat, 7 May 2022 13:10:48 +0200
Subject: Use the user shell by default

Now, by default, it will use the shell configured at /etc/passwd
instead of the one set in the preferences. But still allows to
bypass it and use an specific one if desired.

Fix https://gitlab.com/rastersoft/terminus/-/issues/20
---
 data/interface/properties.ui             | 26 +++++++++++++++++++++-----
 data/org.rastersoft.terminus.gschema.xml |  5 +++++
 meson_scripts/install_schemas.sh         |  2 +-
 src/settings.vala                        | 13 +++++++++++--
 src/terminal.vala                        | 19 ++++++++++++++++++-
 5 files changed, 56 insertions(+), 9 deletions(-)

diff --git a/data/interface/properties.ui b/data/interface/properties.ui
index c5c8e9b..14c94c7 100644
--- a/data/interface/properties.ui
+++ b/data/interface/properties.ui
@@ -53,7 +53,7 @@
         <property name="can-focus">True</property>
         <property name="scrollable">True</property>
         <child>
-          <!-- n-columns=2 n-rows=5 -->
+          <!-- n-columns=2 n-rows=6 -->
           <object class="GtkGrid">
             <property name="visible">True</property>
             <property name="can-focus">False</property>
@@ -87,7 +87,7 @@
               </object>
               <packing>
                 <property name="left-attach">0</property>
-                <property name="top-attach">4</property>
+                <property name="top-attach">5</property>
                 <property name="width">2</property>
               </packing>
             </child>
@@ -135,7 +135,7 @@
               </object>
               <packing>
                 <property name="left-attach">0</property>
-                <property name="top-attach">3</property>
+                <property name="top-attach">4</property>
                 <property name="width">2</property>
               </packing>
             </child>
@@ -158,11 +158,11 @@
                 <property name="visible">True</property>
                 <property name="can-focus">False</property>
                 <property name="halign">end</property>
-                <property name="label" translatable="yes">Command shell:</property>
+                <property name="label" translatable="yes">Custom command shell:</property>
               </object>
               <packing>
                 <property name="left-attach">0</property>
-                <property name="top-attach">2</property>
+                <property name="top-attach">3</property>
               </packing>
             </child>
             <child>
@@ -172,7 +172,23 @@
               </object>
               <packing>
                 <property name="left-attach">1</property>
+                <property name="top-attach">3</property>
+              </packing>
+            </child>
+            <child>
+              <object class="GtkCheckButton" id="use_custom_shell">
+                <property name="label" translatable="yes">Use custom shell</property>
+                <property name="visible">True</property>
+                <property name="can-focus">True</property>
+                <property name="receives-default">False</property>
+                <property name="halign">start</property>
+                <property name="margin-top">8</property>
+                <property name="draw-indicator">True</property>
+              </object>
+              <packing>
+                <property name="left-attach">0</property>
                 <property name="top-attach">2</property>
+                <property name="width">2</property>
               </packing>
             </child>
           </object>
diff --git a/data/org.rastersoft.terminus.gschema.xml b/data/org.rastersoft.terminus.gschema.xml
index 33f60d7..0314133 100644
--- a/data/org.rastersoft.terminus.gschema.xml
+++ b/data/org.rastersoft.terminus.gschema.xml
@@ -105,6 +105,11 @@
       <default>true</default>
       <summary>Whether the terminal should rewrap the text when the window is resized</summary>
       <description></description>
+    </key>
+  <key type="b" name="use-custom-shell">
+      <default>false</default>
+      <summary>Use custom shell</summary>
+      <description></description>
     </key>
 	<!--<key type="" name="">
       <default></default>
diff --git a/meson_scripts/install_schemas.sh b/meson_scripts/install_schemas.sh
index c8885f4..2e769fc 100755
--- a/meson_scripts/install_schemas.sh
+++ b/meson_scripts/install_schemas.sh
@@ -2,5 +2,5 @@
 
 if [[ -z "${DESTDIR}" ]]; then
     echo Compiling gsettings schemas...
-	glib-compile-schemas ${MESON_INSTALL_PREFIX}/share/glib-2.0/schemas
+    glib-compile-schemas ${MESON_INSTALL_PREFIX}/share/glib-2.0/schemas
 fi
diff --git a/src/settings.vala b/src/settings.vala
index ed39e18..b7a64cb 100644
--- a/src/settings.vala
+++ b/src/settings.vala
@@ -26,6 +26,7 @@ namespace Terminus {
 		private Gtk.CheckButton use_bold_color;
 		private Gtk.CheckButton use_cursor_color;
 		private Gtk.CheckButton use_highlight_color;
+		private Gtk.CheckButton use_custom_shell;
 		private Gtk.SpinButton scroll_value;
 		private Gtk.Button custom_font;
 		private Gtk.ColorButton fg_color;
@@ -42,6 +43,7 @@ namespace Terminus {
 		private Gtk.ListStore palette_schemes;
 		private Gtk.ComboBox cursor_shape;
 		private Gtk.ListStore keybindings;
+		private Gtk.Entry custom_shell;
 
 
 		private bool editing_keybind;
@@ -68,6 +70,10 @@ namespace Terminus {
 			label_version.label = _("Version %s").printf(Constants.VERSION);
 
 			this.use_system_font = main_window.get_object("use_system_font") as Gtk.CheckButton;
+			this.use_custom_shell = main_window.get_object("use_custom_shell") as Gtk.CheckButton;
+			use_custom_shell.toggled.connect(() => {
+				this.custom_shell.sensitive = this.use_custom_shell.active;
+			});
 			this.custom_font     = main_window.get_object("custom_font") as Gtk.Button;
 			use_system_font.toggled.connect(() => {
 				this.custom_font.sensitive = this.use_system_font.active;
@@ -240,6 +246,8 @@ namespace Terminus {
 
 			this.enable_guake_mode = main_window.get_object("enable_guake_mode") as Gtk.CheckButton;
 
+			this.custom_shell = main_window.get_object("command_shell") as Gtk.Entry;
+
 			Terminus.settings.bind("cursor-shape", this.cursor_shape, "active", GLib.SettingsBindFlags.DEFAULT);
 			Terminus.settings.bind("use-system-font", this.use_system_font, "active", GLib.SettingsBindFlags.DEFAULT | GLib.SettingsBindFlags.INVERT_BOOLEAN);
 			Terminus.settings.bind("terminal-font", this.custom_font, "font_name", GLib.SettingsBindFlags.DEFAULT);
@@ -249,8 +257,8 @@ namespace Terminus {
 			Terminus.settings.bind("scroll-on-keystroke", main_window.get_object("scroll_on_keystroke") as Gtk.CheckButton, "active", GLib.SettingsBindFlags.DEFAULT);
 			Terminus.settings.bind("enable-guake-mode", this.enable_guake_mode, "active", GLib.SettingsBindFlags.DEFAULT);
 			Terminus.settings.bind("terminal-bell", main_window.get_object("terminal_bell") as Gtk.CheckButton, "active", GLib.SettingsBindFlags.DEFAULT);
-			//Terminus.settings.bind("allow-bold", main_window.get_object("allow_bold") as Gtk.CheckButton, "active", GLib.SettingsBindFlags.DEFAULT);
-			Terminus.settings.bind("shell-command", main_window.get_object("command_shell") as Gtk.Entry, "text", GLib.SettingsBindFlags.DEFAULT);
+			Terminus.settings.bind("shell-command", this.custom_shell, "text", GLib.SettingsBindFlags.DEFAULT);
+			Terminus.settings.bind("use-custom-shell", this.use_custom_shell, "active", GLib.SettingsBindFlags.DEFAULT);
 
 			int counter  = -1;
 			int selected = 0;
@@ -274,6 +282,7 @@ namespace Terminus {
 				selcount++;
 			}
 
+			this.custom_shell.sensitive = this.use_custom_shell.active;
 			this.custom_font.sensitive  = this.use_system_font.active;
 			this.scroll_value.sensitive = !this.infinite_scroll.active;
 
diff --git a/src/terminal.vala b/src/terminal.vala
index fc36384..d7fe88e 100644
--- a/src/terminal.vala
+++ b/src/terminal.vala
@@ -234,7 +234,24 @@ namespace Terminus {
 			newbox.pack_start(right_scroll, false, true);
 
 			string[] cmd = {};
-			cmd += Terminus.settings.get_string("shell-command");
+			if (Terminus.settings.get_boolean("use-custom-shell")) {
+				cmd += Terminus.settings.get_string("shell-command");
+			} else {
+				bool found = false;
+				unowned Posix.Passwd passwd;
+				Posix.setpwent();
+				while (null != (passwd = Posix.getpwent())) {
+					if (passwd.pw_name == GLib.Environment.get_user_name()) {
+						found = true;
+						cmd += passwd.pw_shell;
+						break;
+					}
+				}
+				if (!found) {
+					cmd += "/bin/sh";
+				}
+				Posix.endpwent();
+			}
 			if ((commands != null) && (commands.length != 0)) {
 				cmd += "-c";
 				foreach (var command in commands) {
